### PR TITLE
Update CHANGELOG text structure

### LINKED ISSUE #1935: Non-unique heading anchors on pip-tools' changelog documentation
#### What's the problem this feature will solve?

Currently, pip-tools generates non-unique anchors for links in the documentation, as seen for example here: https://pip-tools.readthedocs.io/en/latest/changelog/#id1. This will led to confusion with links for future versions deployed on RTD.

#### Describe the solution you'd like

Generate unique anchors, for example https://pip-tools.readthedocs.io/en/latest/changelog/#710-2023-07-18

#### Alternative Solutions

N/A

#### Additional context

- Example on GitHub: https://github.com/jazzband/pip-tools/blob/main/CHANGELOG.md#700-2023-07-14
- https://myst-parser.readthedocs.io/en/stable/syntax/optional.html#anchor-slug-structure

### PR DESCRIPTION
**Pull request by atugushev**
Fixes #1935.
Supersedes #1942.
Preview: https://pip-tools--1945.org.readthedocs.build/en/1945/changelog/.

##### Contributor checklist

- [ ] Provided the tests for the changes.
- [x] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,6 @@
-## 7.1.0 (2023-07-18)
```
Comment (webknjaz) [inline CHANGELOG.md:1]: In my initial review, I didn't mean to imply that the Markdown markup needed changes. Instead, I thought, that maybe the anchor transformer would produce anchors without the date segment…

Comment (webknjaz) [reply CHANGELOG.md:1]: Do you think this is what's needed? Will the release bot comply and produce a nicer format?

Comment (atugushev) [reply CHANGELOG.md:1]: > In my initial review, I didn't mean to imply that the Markdown markup needed changes. Instead, I thought, that maybe the anchor transformer would produce anchors without the date segment…

I've found a simpler solution that doesn't involve code changes; less code, fewer bugs. I've just slightly altered the heading format so that Sphinx can successfully generate an ID for it to use as an anchor. Currently, it generates empty string, leading to id1, id2 anchors. For example:

```python
>>> from docutils.nodes import make_id
>>> make_id('7.1.0 (2023-07-18)')
''

>>> make_id('7.1.0')
''

>>> make_id('v7.1.0 (2023-07-18)')
'v7-1-0-2023-07-18'

>>> make_id('v7.1.0')
'v7-1-0'
```

I've also moved the date from the heading and placed it below, similar to what's been done on [cheroot](https://cheroot.cherrypy.dev/en/latest/history/#v9-0-0) and [setuptools](https://setuptools.pypa.io/en/latest/history.html#v68-0-0).

---

> Do you think this is what's needed? Will the release bot comply and produce a nicer format?

Currently, bot does not populate the CHANGELOG itself; it only prepares [release drafts](https://github.com/jazzband/pip-tools/releases). We manually fill the CHANGELOG by copying data from the release draft according to the [contributing guideline](https://pip-tools.readthedocs.io/en/latest/contributing/#project-release-process). In this process, we manually add a heading that includes the version and date.

---

In conclusion, we will simply prepare the changelog in a slightly different format to enhance its appearance on docs. What do you think?

Comment (webknjaz) [reply CHANGELOG.md:1]: Okay, if that works, let's do this. Personally, I'm not a fan of the release drafter automation with all of its quirks, but so long as I don't have to deal with it, I choose not to push for changes in that area.

Comment (webknjaz) [reply CHANGELOG.md:1]: Marking this thread as resolved.

**Review by webknjaz — APPROVED** [review]

**Comment by atugushev** [comment]
Thanks @webknjaz!
